ReSharper.FSharp/src/FSharp.Common/src/Shim/TypeProviders/TypeProvidersManager.fs

```diff
@@ -13,45 +17,51 @@ open JetBrains.ReSharper.Plugins.FSharp.TypeProviders.Protocol.Models
     let typeProvidersPerAssembly = ConcurrentDictionary<_, ConcurrentDictionary<_, IProxyTypeProvider>>()
     let proxyTypeProvidersPerId = ConcurrentDictionary<_, _>()
 
-    let rec addTypeProvider envKey (tp: IProxyTypeProvider) =
-        let fullName = tp.GetDisplayName(fullName = true)
+    let rec addTypeProvider projectAssembly tpAssembly (tp: IProxyTypeProvider) =
+        let tpKey = struct(tpAssembly, tp.GetDisplayName(fullName = true))
 
-        match typeProvidersPerAssembly.TryGetValue(envKey) with
+        match typeProvidersPerAssembly.TryGetValue(projectAssembly) with
         | true, assemblyCache ->
-            match assemblyCache.TryGetValue(fullName) with
+            match assemblyCache.TryGetValue(tpKey) with
             | true, oldTp ->
                 oldTp.Dispose()
-                addTypeProvider envKey tp
+                addTypeProvider projectAssembly tpAssembly tp
             | false, _ ->
-                assemblyCache.TryAdd(fullName, tp) |> ignore
+                assemblyCache.TryAdd(tpKey, tp) |> ignore
                 proxyTypeProvidersPerId.TryAdd(tp.EntityId, tp) |> ignore
-                tp.Disposed.Add(fun _ -> removeTypeProvider envKey tp)
+                tp.Disposed.Add(fun _ -> removeTypeProvider projectAssembly tpKey tp.EntityId)
         | false, _ ->
-            typeProvidersPerAssembly.TryAdd(envKey, ConcurrentDictionary()) |> ignore
-            addTypeProvider envKey tp
+            typeProvidersPerAssembly.TryAdd(projectAssembly, ConcurrentDictionary()) |> ignore
+            addTypeProvider projectAssembly tpAssembly tp
 
-    and removeTypeProvider envKey (tp: IProxyTypeProvider) =
+    and removeTypeProvider projectAssembly tpKey tpId =
         // Removes types in a unified manner, may also be disposed by FCS.
-        typeProvidersPerAssembly.[envKey].TryRemove(tp.GetDisplayName(true)) |> ignore
-        proxyTypeProvidersPerId.TryRemove(tp.EntityId) |> ignore
+        typeProvidersPerAssembly.[projectAssembly].TryRemove(tpKey) |> ignore
+        proxyTypeProvidersPerId.TryRemove(tpId) |> ignore
 
-        if typeProvidersPerAssembly.[envKey].Count = 0 then
-            typeProvidersPerAssembly.TryRemove(envKey) |> ignore
+        if typeProvidersPerAssembly.[projectAssembly].Count = 0 then
+            typeProvidersPerAssembly.TryRemove(projectAssembly) |> ignore
 
-    member x.Add(envKey, tp) =
-        addTypeProvider envKey tp
+    member x.Add(projectAssembly, tpAssembly, tp) =
+        addTypeProvider projectAssembly tpAssembly tp
 
     member x.Get(id) =
         let hasValue = SpinWait.SpinUntil((fun () -> proxyTypeProvidersPerId.ContainsKey id), 15_000)
 
         if not hasValue then failwith $"Cannot get type provider {id} from TypeProvidersCache"
         else proxyTypeProvidersPerId.[id]
 
+    member x.Get(outputPath) =
+        match typeProvidersPerAssembly.TryGetValue(outputPath) with
+        | true, x -> x.Values
+        | false, _ -> [||] :> _
```

Comment:
`[||]` is not currently a singleton and will needlessly allocate. Let's use the empty array instance from `JetBrains.Util` or at least BCL?